litestar/handlers/websocket_handlers/_utils.py

```diff
@@ -0,0 +1,184 @@
+    )
+
+    can_send_data: bool
+    handler_function: AnyCallable
+    json_encoder: JsonEncoder
+    listener_callback: AnyCallable
+    listener_callback_signature: ParsedSignature
+    pass_socket: bool
+    should_encode_to_json: bool
+    wants_receive_type: type
+    resolved_data_dto: type[DTOInterface] | None
+    resolved_return_dto: type[DTOInterface] | None
+    resolved_type_encoders: TypeEncodersMap
+
+
+def _update_listener_fn_signature(listener_context: _ListenerContext) -> None:
+    # make our listener_fn look like the callback, so we get a correct signature model
+
+    callback_signature = listener_context.listener_callback_signature.original_signature
+    new_params = [p for p in callback_signature.parameters.values() if p.name not in {"data"}]
+    if "socket" not in callback_signature.parameters:
+        new_params.append(inspect.Parameter(name="socket", kind=inspect.Parameter.KEYWORD_ONLY, annotation="WebSocket"))
+    # TODO: could this be avoided if logic for construction of a signature model were moved on to the route handler?
+    new_signature = callback_signature.replace(parameters=new_params)
+    listener_context.handler_function.__signature__ = new_signature  # type: ignore[attr-defined]
+    listener_context.handler_function.__annotations__ = {
+        p.name: p.annotation for p in new_signature.parameters.values()
+    }
+
+
+def _create_handler_function(
+    listener_context: _ListenerContext,
+    receive_mode: WebSocketMode,
+    send_mode: WebSocketMode,
+    on_accept: AsyncCallable | None,
+    on_disconnect: AsyncCallable | None,
+) -> Callable[..., Coroutine[None, None, None]]:
+    handle_receive = _create_handle_receive(listener_context, receive_mode)
+    handle_send = _create_handle_send(listener_context, send_mode)
+    return _create_listener_function(
+        handle_receive=handle_receive,
+        handle_send=handle_send,
+        listener_context=listener_context,
+        on_accept=on_accept,
+        on_disconnect=on_disconnect,
+    )
+
+
+def _create_handle_receive(
+    listener_context: _ListenerContext, receive_mode: WebSocketMode
+) -> Callable[[WebSocket, DTOInterface | None], Coroutine[Any, None, None]]:
+    async def handle_receive(socket: WebSocket, dto: DTOInterface | None) -> Any:
+        received_data: Any = await socket.receive_data(mode=receive_mode)  # pyright: ignore
+        if dto:
+            if isinstance(received_data, str):
+                received_data = received_data.encode("utf-8")
+            received_data = dto.bytes_to_data_type(received_data)
+        elif listener_context.wants_receive_type is str:
+            if isinstance(received_data, bytes):
+                received_data = received_data.decode("utf-8")
+        elif listener_context.wants_receive_type is bytes:
+            if isinstance(received_data, str):
+                received_data = received_data.encode("utf-8")
+        else:
+            received_data = decode_json(received_data)
```

Comment:
To optimize this, instead of putting the outer `if`/`elif` tree inside `handle_receive`, `_create_handle_receive` could return a `handle_receive` function specific to the config, e.g. one for DTOs, one for strings etc. Kind of a micro-optimization but I don't see a reason not to do it. Wdyt?

Reply:
I've done this. It means that we need `_ListenerContext.handle_receive` and `_ListenerContext.handle_send` so that we can build them in `on_registration()`, as the conditional needs resolved data.

LMK what you think.

Reply:
It does make things a bit more opaque but I guess it's a solid approach. I'm not 100% sure if the added complexity is worth it, but I'd be okay with having it like this. We can always revisit this should a better pattern for this emerge.